File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/TypeTemplateSelector.vb
Imports Windows.UI.Xaml.Markup

<ContentProperty(Name:=NameOf(TypeTemplateSelector.Selections))>
Public Class TypeTemplateSelector
    Inherits DataTemplateSelector

    Public ReadOnly Property Selections As New List(Of TypeTemplateSelection)
    Protected Overrides Function SelectTemplateCore(item As Object, container As DependencyObject) As DataTemplate
        Return SelectTemplateCore(item)
    End Function
    Protected Overrides Function SelectTemplateCore(item As Object) As DataTemplate
        If item Is Nothing Then Return Nothing
        Dim type = item.GetType()
        For Each s In Selections
            If s.Type = type Then Return s.Template
        Next
        Return Nothing
    End Function
End Class

Public Class TypeTemplateSelection
    Public Property Type As Type
    Public Property Template As DataTemplate
End Class

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/TypeTemplateSelector.cs
using System.Collections.Generic;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Markup;

namespace Sakuno.ING.Views.UWP.Controls
{
    [ContentProperty(Name = nameof(Selections))]
    public class TypeTemplateSelector : DataTemplateSelector
    {
        public List<TypeTemplateSelection> Selections { get; } = new List<TypeTemplateSelection>();
        public DataTemplate Fallback { get; set; }

        protected override DataTemplate SelectTemplateCore(object item)
        {
            if (item is null)
                return null;
            var typeName = item.GetType().Name;
            foreach (var s in Selections)
                if (s.TypeName == typeName)
                    return s.Template;
            return Fallback;
        }

        protected override DataTemplate SelectTemplateCore(object item, DependencyObject container)
            => SelectTemplateCore(item);
    }

    [ContentProperty(Name = nameof(Template))]
    public class TypeTemplateSelection
    {
        // Using `Type` works in coreclr, but not .Net native
        public string TypeName { get; set; }
        public DataTemplate Template { get; set; }
    }
}
